src/index.ts

```diff
@@ -0,0 +1,384 @@
+            }
+            else if (b === RESPONSE_OUTPUT_END) {
+              resolve(this[train_symbol][0][4]);
+              this[train_symbol].shift();
+              this[state_symbol] = STATE_IDLE;
+              return this.next();
+            }
+          }
+          this[train_symbol][0][3] = false; // esc byte
+          this[train_symbol][0][4] += String.fromCharCode(b);
+        }
+      }
+    }
+  }
+
+  serial_watchdog() {
+    if (!this[train_symbol][0])
+      return;
+    var [, , reject] = this[train_symbol][0]!;
+    this[train_symbol].shift();
+    reject(new Error('ERROR_WATCHDOG'));
+    this[state_symbol] = STATE_IDLE;
+    this.next();
+  }
+
+  open(port?: string, options?: RyderSerialOptions): void {
+    this.options.debug && console.debug('ryderserial attempt open');
+    this.closing = false;
+    if (this.serial && this.serial.isOpen)
+      return;
+    if (this.serial)
+      this.close();
+    this.port = port || this.port;
+    this.options = options || this.options || {};
+    if (!this.options.baudRate)
+      this.options.baudRate = 115200;
+    if (!this.options.lock)
+      this.options.lock = true;
+    if (!this.options.reconnectTime)
+      this.options.reconnectTime = 1000;
+    this.serial = new SerialPort(this.port, this.options);
+    this.serial.on('data', this.serial_data.bind(this));
+    this.serial.on('error', error => {
+      console.log(`this.serial encountered an error: ${error}`)
+      if (this.serial && !this.serial.isOpen) {
+        clearInterval(this[reconnect_symbol]);
+        // TODO: confirm that setInterval is using Node's implementation; and if it's not WHY
+        this[reconnect_symbol] = setInterval(this.open, this.options.reconnectTime);
+        this.emit('failed', error);
+      }
+      this.serial_error.bind(this);
+    });
+    this.serial.on('close', () => {
+      this.options.debug && console.debug('ryderserial close');
+      this.emit('close');
+      clearInterval(this[reconnect_symbol]);
+      if (!this.closing)
+        this[reconnect_symbol] = setInterval(this.open.bind(this), this.options.reconnectTime);
+    });
+    this.serial.on('open', () => {
+      this.options.debug && console.debug('ryderserial open');
+      // reset or keep?
+      // this[train_symbol] = [];
+      // this[state_symbol] = STATE_IDLE;
+      // this[lock_symbol] = [];
+      clearInterval(this[reconnect_symbol]);
+      this.emit('open');
+      this.next();
+    });
+  };
+
+  close(): void {
+    if (this.closing)
+      return;
+    this.closing = true;
+    this.clear();
+    this.serial?.close();
+    clearInterval(this[reconnect_symbol]);
+    this.serial = undefined;
+  };
+
+  locked(): boolean {
+    return !!this[lock_symbol].length;
+  };
+
+  lock(): Promise<void> {
+    this.options.debug && console.debug('ryderserial lock');
+    if (!this[lock_symbol].length) {
+      // TODO: fix this[lock_symbol]... why is it holding 'false'?
```

Comment:
Just to not construct a pointless Promise. Then `unlock()` can be refactored too.

Reply:
Gotcha. Well, it breaks type for `this[lock_symbol]` if ever it holds boolean. 

What is the circumstance that we'd want the `false` in there, beyond avoiding the construction of a pointless Promise? It looks like we're only ever pushing that false value (or pointless promise) if we call `this.lock()` and `this[lock_symbol]` is empty, but it's a bit unclear to me as to **why**. 

In `unlock`, we call 

```typescript
var resolve = this[lock_symbol].shift();
if (!resolve && this[lock_symbol].length) {
    resolve = this[lock_symbol].shift();
} 
resolve && resolve();
```

Which is acting as if we want to waste that first lock.

If I called
```typescript
const ryder_serial = new RyderSerial(/*port, options*/);
ryder_serial.lock(); // first time locking
// this[lock_symbol].push(false);
// return Promise.resolve();
ryder_serial.unlock(); 
// var resolve =  this[lock_symbol].shift(); // --- resolve now equals `false`, this[lock_symbol] is now empty
// resolve && resolve(); // clearly won't call, but why would we need to call it anyway?
```

I guess I just don't fully understand why `this[lock_symbol]` is holding promises at all. It seems like we use `this[lock_symbol]` as an array of arbitrary placeholders that symbolize a "lock" and simply push and pop them throughout. If it's empty? RyderSerial is unlocked! If it's not, RyderSerial is locked. 

Is there a reason that we're not simply using a `this.lock_count`?

Reply:
So developers can do this:

```JS
async () =>
 {
 await serial.lock(); // wait until we get a lock
 // do async stuff, other places that requested
 // a lock later will wait until we unlock.
 serial.unlock();
 }
 ```

If there are no pending lock then it resolves right away but we would still like to know a lock was requested, hence the `false`. I just added a falsy value. We can put in a `Promise.resolve` too and simplify `unlock()`.

Check out `RyderSerial.prototype.sequence()`.

Reply:
Understood. That makes perfect sense. Thanks for clarifying.

Can we keep the pointless promise for now for sake of type simplification? The boolean was a beast to manage with TS.

If that's fine I'll go ahead and update `unlock` as well.